pandas/util/_decorators.py

```diff
@@ -247,6 +247,36 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
     return decorate
 
 
+def doc(*args: Union[str, Callable], **kwargs: str) -> Callable:
+    """
+    A decorator take docstring templates, concatenate them and perform string
+    substitution on it.
+
+    This decorator should be robust even if func.__doc__ is None.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs) -> Callable:
+            return func(*args, **kwargs)
+
+        templates = [func.__doc__ if func.__doc__ else ""]
+        for arg in args:
+            if isinstance(arg, str):
+                templates.append(arg)
+            elif hasattr(arg, "_docstr_template"):
+                templates.append(arg._docstr_template)  # type: ignore
```

Comment:
What are all of the type: ignore comments for?

Reply:
It seems mypy didn't let us declare and use attributes of function objects. They have a [issue](https://github.com/python/mypy/issues/2087) for this, but seems there are no official solutions. 

We might be able to use some trick to handle this, but I am not sure if it worth. It will introducing more "unrelated" code. Also, this call of `_docstr_template` has been checked by `hasattr`, so I think it is safe here.

Just to be clear, I am open to make the adjustment to avoid this ignore comment.

Reply:
I think just adding `https://github.com/python/mypy/issues/2087` as a comment on a line preceding the ignores would be fine.